backstory/src/test/java/com/google/sps/story/DatamuseRequestClientTest.java

```diff
@@ -0,0 +1,339 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.story;
+
+import static org.mockito.Mockito.*;
+
+import au.com.origma.perspectiveapi.v1alpha1.models.AttributeType;
+import com.google.gson.Gson;
+import com.google.sps.APINotAvailableException;
+import com.google.sps.story.data.HttpUrlStreamHandler;
+import com.google.sps.story.data.DatamuseRequestClient;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.net.MalformedURLException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.net.URLStreamHandlerFactory;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.ArgumentMatchers;
+import org.mockito.BDDMockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+/** Quality tests for DatamuseRequestClient */
+@RunWith(JUnit4.class)
+public final class DatamuseRequestClientTest {
+
+  /** a default url to test with */
+  private static final String DEFAULT_URL = "http://sample-website.com/fake?";
+
+  /** the stream handlet to use for tests in this class (necessary to mock URLConnection) */
+  private static HttpUrlStreamHandler httpUrlStreamHandler;
+ 
+  /** 
+   * Set up the URLStreamHandlerFactory with the HTTPUrlStreamHandler 
+   * in order to mock URLConnections to properly test methods 
+   * that call outside URLs. Code comes from this tutorial: 
+   * https://claritysoftware.co.uk/mocking-javas-url-with-mockito/.
+   */
+  @BeforeClass
+  public static void setUpURLStreamHandlerFactory() {
+    // Allows for mocking URL connections
+    URLStreamHandlerFactory urlStreamHandlerFactory = mock(URLStreamHandlerFactory.class);
+    URL.setURLStreamHandlerFactory(urlStreamHandlerFactory);
+ 
+    httpUrlStreamHandler = new HttpUrlStreamHandler();
+    BDDMockito.given(urlStreamHandlerFactory.createURLStreamHandler("http")).willReturn(httpUrlStreamHandler);
+  }
+
+  /**
+   * Reset the URLConnections of this stream handler before every method,
+   * so that URLConnections injected for other tests do not mess 
+   * with the current test. Code taken from this tutorial: 
+   * https://claritysoftware.co.uk/mocking-javas-url-with-mockito/.
+   */
+  @Before
+  public void reset() {
+    httpUrlStreamHandler.resetConnections();
+  }
```

Comment:
Seems like this could be avoided by making `httpUrlStreamHandler` non-static.

Reply:
Refer to above comment. Please let me know if you still want me to change this.

Reply:
Yeah, it's possible that there's other state that's carried over (if not now, then if HttpUrlStreamHandler adds new functionality in the future)... it would be safer to start each test with as clean a slate as possible.

Reply:
I do agree, but unfortunately it looks like this choice is the only option because httpUrlStreamHandler seems like it can only be set once in the class. Please let me know if you see another option.

Reply:
Done

Reply:
That makes sense. For the sake of time, I'll keep this static, but this is good to keep in mind. Thanks!